### ISSUE TITLE
3.29-12: cannot open file by double click in OS X

### ISSUE THREAD

**Issue by teobaldo3524** [fesch/Structorizer.Desktop#744]
Double-clicking a file does not open it: instead, 3.29-12 opens with a blank diagram.
I have to use File> Open or Open recent…

Problem does NOT occur in 3.29-02.

**Comment by codemanyak** [comment]
Which way did you install Structorizer 3.29-12, and on which operating system? (This is not quite irrelevant for your issue.)

**Comment by codemanyak** [comment]
@teobaldo3524 
I wrote:
> Which way did you install Structorizer 3.29-12, and on which operating system?

Okay, I just learned from #743 that you work on macOS. In case you installed two versions of Structorizer in parallel, there might be some OS irritation about what product version to launch (just a guess). After a complete installation, however, you ought to expect that the newer version should be invoked...

**Comment by teobaldo3524** [comment]
"complete installation"?
There's no Installer - you just drag the application to where you want it. That's all.

Yes, the system knows to launch the newer version when a nsd file is double-clicked.
Screenshot of context menu "open with" shows 3.29-12 as default (standaard):



And I dug this up from an old changelog:
Version 3.16 (2009-08-22)
- MAC: Files do not open with double click (bug #7 <https://structorizer.fisch.lu/index.php?include=bugs&bug=7>) [Thijs Zandwijk]
That's me, yes. I wouldn't know if this is the same problem.

To make sure I removed the older version, and dragged ~/.structorizer/ to the desktop.
That made no difference.

**Comment by codemanyak** [comment]
We changed indeed the OSX adapter with version 3.29-12, so it is possible (or even plausible) that the "openFileHandler" hadn't been properly configured with the current version. We will investigate this.

**Comment by codemanyak** [comment]
@fesch 
My guess is that the defect might my caused by the lacking file handler specification in `Mainform.doOSX()` (line is outcommented):
```java
    public void doOSX() {
        try {
            // Generate and register the OSXAdapter, passing it a hash of all the methods we wish to
            // use as delegates for various com.apple.eawt.ApplicationListener methods
            OSXAdapter.setQuitHandler(this, getClass().getDeclaredMethod("quit", (Class[]) null));
            OSXAdapter.setAboutHandler(this, getClass().getDeclaredMethod("about", (Class[]) null));
            OSXAdapter.setPreferencesHandler(this, getClass().getDeclaredMethod("preferences", (Class[]) null));
            OSXAdapter.setDockIconImage(getIconImage());
            //OSXAdapter.setFileHandler(this, getClass().getDeclaredMethod("loadImageFile", new Class[]{String.class}));

        } catch (Exception e) {
            //System.err.println("Error while loading the OSXAdapter:");
            e.printStackTrace();
        }
    }
```
If my assumption should turn out to be true (couldn't test it) then the problem might perhaps be solvable by simply defining a method `Mainform.loadFile()` als follows:
```java
    // General file handler; fed to the OSXAdapter as the method to call when 
    // a file associated to Structorizer is double-clicked or dragged onto it:
    public void loadFile(String filePath) {
        diagram.openNsdOrArr(filePath);
    }
```
... and by activating the outcommented line in `Mainform.doOSX()` with the following change (handler method name):
```java
            OSXAdapter.setFileHandler(this, getClass().getDeclaredMethod("loadFile", new Class[]{String.class}));
```

**Comment by codemanyak** [comment]
@teobaldo3524 
Perhaps we are looking with a wrong scope or in a wrong direction. If I get it right, the last Structoizer version you compare with was 3.29-02. So possibly 3.29-12 is not the first version that fails to open a double-clicked file? What about versions 3.29-03 through 3.29-11?
While Bob doesn't seem available for tests in a Mac environment you might possibly check which of the intermediate releases actually made the difference (you find their downloadable packages among the [releases](https://github.com/fesch/Structorizer.Desktop/releases))?
Thanks in advance for helping us to tackle the problem.

**Comment by teobaldo3524** [comment]
Will do.

Does dropping a file on the application icon trigger the same mechanism as double-clicking?

**Comment by fesch** [comment]
@teobaldo3524 : could you please try the newly uploaded version on the server?

**Comment by teobaldo3524** [comment]
@fesch: I'm unsure about what it is you want me to do.

@ codemanyak: I found that 3.29-10 is the last version to perform open-on-double-click.

**Comment by codemanyak** [comment]
@teobaldo3524 
> @fesch: I'm unsure about what it is you want me to do.

Bob did something "abominable": He stealthly modified the downloadable release according to my fixing proposal, still will the same version label: So what you can download currently as latest version slightly differs from what was available before. Bob just asks you whether this new package fixes your problem.

> @codemanyak: I found that 3.29-10 is the last version to perform open-on-double-click.

Okay, this corroborates my assumption that is has something to do with the new OSX Adapter. So I really wonder how launching an application by double-clickig a file of an associated type is doe in OS X.

**Comment by codemanyak** [comment]
> Does dropping a file on the application icon trigger the same mechanism as double-clicking?

Not with Windows or Linux - they use completely different mechanisms: As outlined in a comment before, double-clicking is usally addressed by starting a new application process fed with the file paths as command-line arguments whereas dragging uses a specific file drop mechanism. I have no idea at all how handles this. The integration of some application functions in the OS X system menu suggests a different philosophy. If OS X avoids to open several process instances then indeed double-clicking and dragging may possibly be handled the same way via a system event sent to the (already running) application. No idea how OS X ensures that the event isn't missed while the process is starting the first time.

**Comment by teobaldo3524** [comment]
Bob just asks you whether this new package fixes your problem.

Ah. Well, it didn't.

**Comment by codemanyak** [comment]
That's too bad. I see two directions of analysis (within an OS X system, i.e. an `com.apple.eawt.Application` environment):
- Either the mechanism doesn't work at all (though it seems to have worked with a slightly different adapter)
- or the openFile event slips through before the diagram is established.

If the latter case can be verified then we might try to cache and postpone the event.

**Comment by teobaldo3524** [comment]
Versions that don't open files on double-click do not open them on drop either.

**Comment by codemanyak** [comment]
So, eventually I got hold of a Macbook of a colleague. My test with a fresh installation of the (improved) version 3.29-12 showed the following behaviour:
1. Double-clicking an nsd file opens Stuctorizer but fails to load the diagram.
1. Double-clicking an nsd when Structorizer already runs works.
1. Dragging an nsd file into an open Structorizer window also works.

My conclusion is: The failure on startup seems to be a concurrency problem indeed - apparently the event is triggered before Structorizer gets ready for input. So I will implement a mechanism to queue the premature file events and to serve them when Structorizer initialization is done.

@teobaldo3524 
Before I do this I would like to ask you to confirm the above behaviour (particularly 2. and 3.) on your computer. Please double-click the nsd file again after Structorizer has opened - then the diagram ought to show up there (as a temporary workaround). 
Thanks in advance.

**Comment by teobaldo3524** [comment]
Yes, that's how it works over here as well.

At the first attempt double-clicking an  NSD did open Structorizer and the diagram, but this was just once.
Seems to confirm it is a timing issue.

**Comment by codemanyak** [comment]
@teobaldo3524 
>Seems to confirm it is a timing issue.

Excellent. Thanks a lot! Then I will run some final tests on a Macbook and prepare a new release if these work.

**Comment by codemanyak** [comment]
My changes did NOT work. I set up a queue for file names possibly notified prematurely, I even moved the handler configuration to the beginning of the constructor lest an event should be missed during a long lasting initialization phase.
It it should not even be necessary if I beleave what is told in the Stackoverflow thread [Double click document file in Mac OS X to open Java application](https://stackoverflow.com/questions/1575190/double-click-document-file-in-mac-os-x-to-open-java-application) (from which I also leaned that the file type association keys we use in the Info.plist are decprecated and that the os.name property test may have to be loosened, too, to detect all OS X versions properly): The file open events are said to be queued and mysteriously postponed to the end of the main function by the system.
So I will have to try again with more logging as soon as I get access to that Macbook again.

For the meantime: Just double-click twice if Structorizer isn't open. :unamused:

**Comment by codemanyak** [comment]
In the event it was so trivial: The file handler must just be the first one of the handlers to be assigned.

**Comment by codemanyak** [comment]
Remark. In https://github.com/fesch/Structorizer.Desktop/issues/744#issuecomment-532582885 I wrote
> (from which I also learned [...] that the os.name property test may have to be loosened, too, to detect all OS X versions properly)

The mentioned [StackOverflow posting](https://stackoverflow.com/questions/1575190/double-click-document-file-in-mac-os-x-to-open-java-application) referred to version 10.8, whereas I tested with Mojave (10.14) successfully using the old os.name detection, so the warning might have been obsolete.

**Comment by codemanyak** [comment]
@teobaldo3524 
As Bob seems to be absent or too occupied to release the new version, you might download the pre-release here.
[structorizer_3.29-13pre_mac.zip](https://github.com/fesch/Structorizer.Desktop/files/3636800/structorizer_3.29-13pre_mac.zip)
